GVFS/GVFS.Virtualization/Projection/GitIndexProjection.cs

```diff
@@ -1322,7 +1391,42 @@ private void UpdatePlaceholders()
                 long millisecondsUpdatingFolderPlaceholders = stopwatch.ElapsedMilliseconds;
 
                 stopwatch.Restart();
+                DateTime updateTime = DateTime.Now;
+                int folderPlaceholdersRefreshAttempts = 0;
+                int folderPlaceholdersRefreshed = 0;
+                foreach (string folderPlaceholderPath in folderPlaceholdersToKeep)

```

Comment:
This will only refresh folders that are placeholders right?  Do we need to do anything for full folders that change or have changes in them?

Reply:
> Do we need to do anything for full folders that change or have changes in them?

I don't believe that we'll need to do anything.  

The behavior of Explorer with full folders will be the same as the behavior of a regular vanilla git repo in Explorer.  I think that *should* be OK because the problem noted in #1532 was with changes that don't result in anything actually changing on disk:

>The issue where Explorer doesn’t refresh when folders are removed from the projection sounds expected. Explorer keeps an open handle to the folder it is showing. I believe it uses directory change notifications to tell it to refresh the view. Altering the projection doesn’t fire a change notification because it doesn’t cause anything in the open folder to be changed, not for real anyway. 

I can test out this scenario a bit, but as long full folder behavior is consistent with native git I think that's sufficient.

Reply:
@kewillford, I just tried testing this out, and found that I can't prune children of full folders.  Is that expected?

(In this test `GVFS\GVFS.FunctionalTests` is full)

```
M:\testSparse\ForTests>gvfs sparse --set GVFS\GVFS.FunctionalTests\Should --prune
Running git status...Succeeded
Updating sparse folder set...Succeeded
Forcing a projection change...Succeeded
Finding folders to prune...Succeeded
Found 5 folders to prune.
Running git status before prune to make sure you don't have any pending changes.

Running git status...Succeeded

WARNING: If you abort the sparse after this point, the repo may become corrupt

Unmounting...Succeeded
Cleaning up folders...-Cannot prune folder 'GVFS\GVFS.FunctionalTests\FileSystemRunners': Must prune parent folder 'GVFS/GVFS.FunctionalTests/'.
Cannot prune folder 'GVFS\GVFS.FunctionalTests\Properties': Must prune parent folder 'GVFS/GVFS.FunctionalTests/'.
Cannot prune folder 'GVFS\GVFS.FunctionalTests\TestData': Must prune parent folder 'GVFS/GVFS.FunctionalTests/'.
Cannot prune folder 'GVFS\GVFS.FunctionalTests\Tests': Must prune parent folder 'GVFS/GVFS.FunctionalTests/'.
Cannot prune folder 'GVFS\GVFS.FunctionalTests\Tools': Must prune parent folder 'GVFS/GVFS.FunctionalTests/'.
Cleaning up folders...Succeeded
Mounting...Succeeded
```

I *did* find a problematic scenario while doing this testing though:

1. `gvfs sparse --set GVFS\GVFS.FunctionalTests --prune`
2. `mkdir GVFS\FullFolder`
3. `gvfs sparse --set GVFS\GVFS;GVFS\FullFolder --prune`
4. `gvfs sparse --set GVFS\GVFS.FunctionalTests --prune`

After step 4 Explorer will refresh.

It looks like the issue is that the `GVFS` directory is not refreshed after the unmount\dehydrate\mount.

This is not an issue in the root directory because with the current changes it is always refreshed.

Reply:
@kewillford and I just discussed this offline.  I am going to try moving the refresh logic out of `GitIndexProjection` and into the `sparse` and `dehydrate` verbs.

Advantages:

- `sparse` can perform the refresh after the final `mount`
- No risk of impacting the "update projection" hot path

Reply:
>  I am going to try moving the refresh logic out of GitIndexProjection and into the sparse and dehydrate verbs.

Now that I've taken another look that won't be quite so simple, as we don't always need to run dehydrate.  I will tag this PR as WIP as I dig into some options here.

Reply:
> Now that I've taken another look that won't be quite so simple, as we don't always need to run dehydrate. I will tag this PR as WIP as I dig into some options here.

I actually was able to do this and it's in the latest commit.  @kewillford I have not done any testing of these changes at all yet, and so please feel free to hold off on reviewing.

Reply:
>  @kewillford I have not done any testing of these changes at all yet, and so please feel free to hold off on reviewing.

@kewillford I did some tests with the latest changes on my machine and found that adding/removing ~50 L3 folders (each with unique parents and grandparents) took between 20 and 40ms:

*Adding the directories to the projection (3 runs)*
```
[2019-11-15 15:31:54.6444 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":103,"refreshErrors":0,"ElapsedMilliseconds":31,"Message":"Updated folder last write times"}
[2019-11-15 15:32:58.6538 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":103,"refreshErrors":0,"ElapsedMilliseconds":19,"Message":"Updated folder last write times"}
[2019-11-15 15:33:44.5356 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":103,"refreshErrors":0,"ElapsedMilliseconds":19,"Message":"Updated folder last write times"}
```

*Removing the directories to the projection (3 runs)*
```
[2019-11-15 15:32:37.7900 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":1,"refreshErrors":0,"ElapsedMilliseconds":27,"Message":"Updated folder last write times"}
[2019-11-15 15:33:11.2884 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":1,"refreshErrors":0,"ElapsedMilliseconds":37,"Message":"Updated folder last write times"}
[2019-11-15 15:37:36.6737 -08:00] UpdateParentFolderLastWriteTimes_Summary {"foldersToRefresh":103,"foldersRefreshed":1,"refreshErrors":0,"ElapsedMilliseconds":14,"Message":"Updated folder last write times"}
```

And so I am thinking that performance is sufficient for the use cases we're aware of today.

Please let me know what you think of the perf and the latest changes, thanks!